File: count_subarrays_where_max_element_appears_at_least_k_times.cpp
class Solution {
public:
    long long countSubarrays(vector<int>& nums, int k) {
        int maxi = *max_element(nums.begin(), nums.end());
        int n = nums.size(), count = 0;
        long long ans = 0;

        for(int l=0, r=0; r < n; r++) {
            count += (nums[r] == maxi);
            
            while(count >= k) {
                count -= (nums[l++] == maxi);
            }
            ans += l;
        }
        return ans;
    }
};

// Time Complexity : O(n)
// Space Complexity : O(1)

File: count_subarrays_where_max_element_appears_at_least_k_times.go
func countSubarrays(nums []int, k int) int64 {
    n, count, maxi := len(nums), 0, 0
    var ans int64 = 0

    for _, val := range nums {
        maxi = max(maxi, val)
    }

    for l, r := 0, 0; r < n; r++ {
        if nums[r] == maxi {
            count += 1
        } 

        for count >= k {
            if nums[l] == maxi {
                count -= 1
            }
            l++
        }
        ans += int64(l)
    }
    return ans
}

// Time Complexity: O(n)
// Space Complexity: O(1)
